### DISCUSSION TITLE
Tree: Keep tree nodes collapsed after deleting filter

### DISCUSSION THREAD

**Discussion by eftewuer** [primefaces/community#2513] [PrimeFaces] [answered]
Hello

The ```<p:tree>``` component behavior with filtering looks like this:
1. you enter a text in the filter field, press enter and the tree filters the nodes accordingly and expands them so the subnodes are visible
2. then you remove the filter text and press enter - the tree shows all nodes correctly and they are all expanded.

I have a big number of nodes with their subnodes with their subnodes. If a user goes to step 2. - which is deleting the filter text - I would like to display all the nodes collapsed.

Is this supported?

P.S. I tried to used ```<p:tree` filterFunction="#{myBean.filterFunction}">``` and in the bean method call ```"treeNode.setExpanded(false);"``` for every node but it doesn't work. From tree control behavior it looks that first the ```myBean.filterFunction()``` is called and after that some PF mechanism makes the nodes expanded anyway.

Also tried another approach: in ```<p:tree>``` define 
```
<p:ajax event="filter" oncomplete="onFilterComplete();"
//where the onFilterComplete() is a remoteCommand
<p:remoteCommand name="onFilterComplete" action="#{myBean.onFilterComplete}">
```
and in ```myBean.onFilterComplete()``` iterate over children of the root node and call ```node.setExpanded(false);``` . But it also didn't work.

**Comment by melloware** [comment]
I would think `expanded` should be respected?  

Can you provide an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project so we can see your issue?

**Comment by eftewuer** [comment]
Hello

It is possible to test it with PF Showcase example:
http://www.primefaces.org/showcase/ui/data/tree/filter.xhtml?jfwid=e9d99

The example there works the same way:
1. open the page - all tree nodes are already collapsed
2. enter any matching filter in Basic tree or Custom tree on that page (I enter "go" for example) - only filter matching nodes are visible - and that is correct.
3. remove the filter - the trees show all the nodes expanded.

Is it supported to make all the nodes being collapsed right after removing the filter text? So not only to preserve expanded/collapsed state from before the filtering - but to enforce them to be all collapsed.

**Reply by melloware** [reply]
OK that looks like a bug can you open a PrimeFaces bug ticket: https://github.com/primefaces/primefaces/issues

**Reply by eftewuer** [reply, marked as answer]
Thank you for the suggestion. The ticket I've created is https://github.com/primefaces/primefaces/issues/12481
